hazelcast/src/test/java/com/hazelcast/client/ringbuffer/RingbufferTest.java

```diff
@@ -88,27 +89,62 @@ public void tearDown() {
         hazelcastFactory.terminateAll();
     }
 
+    @Test(expected = IllegalArgumentException.class)
+    public void readManyAsync_whenStartSequenceIsNegative() {
+        clientRingbuffer.readManyAsync(-1, 1, 10, null);
+    }
+
     @Test
-    public void readManyAsync_whenHitsStale_shouldNotBeBlocked() throws Exception {
+    public void readManyAsync_whenStartSequenceIsNoLongerAvailable_getsClamped() throws Exception {
+        serverRingbuffer.addAllAsync(asList("0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10"), OVERWRITE);
         CompletionStage<ReadResultSet<String>> f = clientRingbuffer.readManyAsync(0, 1, 10, null);
+
+        ReadResultSet rs = f.toCompletableFuture().get();
+        assertEquals(10, rs.readCount());
+        assertEquals("1", rs.get(0));
+        assertEquals("10", rs.get(9));
+    }
+
+    @Test
+    public void readManyAsync_whenStartSequenceIsEqualToTailSequence() throws Exception {
         serverRingbuffer.addAllAsync(asList("0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10"), OVERWRITE);
-        expectedException.expect(new RootCauseMatcher(StaleSequenceException.class));
-        f.toCompletableFuture().get();
+        CompletionStage<ReadResultSet<String>> f = clientRingbuffer.readManyAsync(10, 1, 10, null);
+
+        ReadResultSet rs = f.toCompletableFuture().get();
+        assertEquals(1, rs.readCount());
+        assertEquals("10", rs.get(0));
+    }
+
+    @Test
+    public void readManyAsync_whenStartSequenceIsJustBeyondTailSequence() throws Exception {
+        serverRingbuffer.addAllAsync(asList("0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10"), OVERWRITE);
+        CompletionStage<ReadResultSet<String>> f = clientRingbuffer.readManyAsync(11, 1, 10, null);
+
+        //test that the read blocks (should at least fail some of the time, if not the case)
+        TimeUnit.MILLISECONDS.sleep(250);
```

Comment:
I find the following better understandable:

```
f.toCompletableFuture().get(250, TimeUnit.MILLISECONDS);
```
And add throws to test method `(expected = TimeoutException.class)`

Reply:
Ok.